internal/util/idlocker.go

```diff
@@ -59,3 +61,202 @@ func (vl *VolumeLocks) Release(volumeID string) {
+		if _, ok := ol.locks[deleteOpt][volumeID]; ok {
+			return fmt.Errorf("a Delete operation with given id %s already exists", volumeID)
+		}
+		// increment the counter for snapshot create operation
+		val := ol.locks[createOpt][volumeID]
+		ol.locks[createOpt][volumeID] = val + 1
+	case cloneOpt:
+		// check any delete operation is going on for given volume ID
+		if _, ok := ol.locks[deleteOpt][volumeID]; ok {
+			return fmt.Errorf("a Delete operation with given id %s already exists", volumeID)
+		}
+
+		// check any expand operation is going on for given volume ID
+		if _, ok := ol.locks[expandOpt][volumeID]; ok {
+			return fmt.Errorf("an Expand operation with given id %s already exists", volumeID)
+		}
+		// increment the counter for clone operation
+		val := ol.locks[cloneOpt][volumeID]
+		ol.locks[cloneOpt][volumeID] = val + 1
+	case deleteOpt:
+		// check any expand operation is going on for given volume ID
+		if _, ok := ol.locks[expandOpt][volumeID]; ok {
+			return fmt.Errorf("an Expand operation with given id %s already exists", volumeID)
+		}
+		// check any clone operation is going on for given volume ID
+		if _, ok := ol.locks[cloneOpt][volumeID]; ok {
+			return fmt.Errorf("a Clone operation with given id %s already exists", volumeID)
+		}
+		// check any delete operation is going on for given volume ID
+		if _, ok := ol.locks[createOpt][volumeID]; ok {
+			return fmt.Errorf("a Create operation with given id %s already exists", volumeID)
+		}
+
+		// check any expand operation is going on for given volume ID
+		if _, ok := ol.locks[expandOpt][volumeID]; ok {
+			return fmt.Errorf("an Expand operation with given id %s already exists", volumeID)
+		}
+		// check any restore operation is going on for given volume ID
+		if _, ok := ol.locks[restoreOpt][volumeID]; ok {
+			return fmt.Errorf("a Restore operation with given id %s already exists", volumeID)
+		}
+		ol.locks[deleteOpt][volumeID] = 1
+	case restoreOpt:
+		// check any delete operation is going on for given volume ID
+		if _, ok := ol.locks[deleteOpt][volumeID]; ok {
+			return fmt.Errorf("a Delete operation with given id %s already exists", volumeID)
+		}
+		// increment the counter for restore operation
+		val := ol.locks[restoreOpt][volumeID]
+		ol.locks[restoreOpt][volumeID] = val + 1
+	case expandOpt:
+		// check any delete operation is going on for given volume ID
+		if _, ok := ol.locks[deleteOpt][volumeID]; ok {
+			return fmt.Errorf("a Delete operation with given id %s already exists", volumeID)
+		}
+		// check any clone operation is going on for given volume ID
+		if _, ok := ol.locks[cloneOpt][volumeID]; ok {
+			return fmt.Errorf("a Clone operation with given id %s already exists", volumeID)
+		}
+		ol.locks[expandOpt][volumeID] = 1
+	default:
+		return fmt.Errorf("%v operation not supported", op)
+	}
+	return nil
+}
+
+// GetSnapshotCreateLock gets the snapshot lock on given volumeID.
+func (ol *OperationLock) GetSnapshotCreateLock(volumeID string) error {
```

Comment:
My initial feeling is that these functions are not very useful. Make the operations like `createOpt` available as `CREATE_OPERATION` and pass it to `TryAcquire()` directly is just as clear (if not clearer).

Or, make `tyrAcquire()` and `release()` internal, and force using these function.